### ISSUE TITLE
Custom javascript injection

### ISSUE THREAD

**Issue by in4u** [uBlockOrigin/uBlock-issues#309]
uBO's quite powerful and efficient as it is. However, in light of some recent filter-related issues, I realized that if support for custom javascript injection was added then there would be no base left uncovered by uBO.

Like for https://github.com/uBlockOrigin/uAssets/issues/4143, Adguard has [this](https://github.com/AdguardTeam/AdguardFilters/commit/c8bad639f7c430be53ad1d21d46debf27df9fcdf#diff-b4a21fbbe41a63b9ce05047244f734e5R94) nifty solution.

And such a feature would have also helped in https://github.com/uBlockOrigin/uAssets/issues/4138#issuecomment-439601926. In the same thread, there was a discussion about implementing something along the lines of ABP's `$rewrite` (useful for certain other issues too) which I feel could probably be redundant if custom javascript injection was supported.

The risk of misuse of this powerful feature can be mitigated to a large extent by allowing it for trusted lists only, similar to Adguard's [implementation](https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters#javascript-rules).

Please note that I tried searching for issues related to this feature enhancement here but couldn't find one. So, my apologies in case this has been discussed before and a decision has already been taken in this regard.

**Comment by uBlock-user** [comment]
Cookie injection is that what you're proposing ?

**Comment by in4u** [comment]
What I proposed includes but is not restricted to cookie injection.

**Comment by uBlock-user** [comment]
I have updated my reply FYI

**Comment by uBlock-user** [comment]
Adguard's nifty solution replies on cookie insertion which is what your proposal is based on and $rewrite will not be implemented as it was already declined before - https://github.com/uBlockOrigin/uBlock-issues/issues/46#issuecomment-391303700

**Comment by in4u** [comment]
I have addressed how to address security concerns in my post. The feature could be limited to trusted lists only, including official uBO lists and user's personal list. If even uBO's lists can't be trusted then IDK how to feel about that. Maybe just allow for user list to begin with?

**Comment by uBlock-user** [comment]
> Maybe just allow for user list to begin with?

Like I said, you could use a custom scriptet and allocate the URL in userResourcesLocation via advancedsettings.html

**Comment by in4u** [comment]
Yes, that might take care of begin with. But what about the next step? Can't uBO's own lists be trusted?!

**Comment by uBlock-user** [comment]
It's not more about uBO's own lists but rather external lists making use of uBO's scriptlets.

**Comment by in4u** [comment]
That is not a problem at all. As I noted earlier, only trusted lists could be allowed to use proposed functionality. How can external lists make use of it if they are not explicitly allowed? Quoting from Adguard's [implementation](https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters#javascript-rules) page:

>Please note that this type of rules can be used **only in trusted filters**. This category includes your own **User filter** and all the filters created by AdGuard Team.

Something similar can be done for uBO where only uBO's official lists are allowed along with user's own.

**Comment by jspenguin2017** [comment]
Privileged scriptlets were discussed before: https://github.com/uBlockOrigin/uAssets/pull/2099

**Comment by in4u** [comment]
I'm not much familiar with uBO's internal scriptlet system and it was only recently that I was introduced to [this](https://github.com/gorhill/uBlock/wiki/Resources-Library) page (which somehow isn't prominently, if at all, linked to in the main wiki). Going through the link shared by @jspenguin2017 gave me an idea, so I am just going to think out loud here:

1. Create a scriptlet `custom.js` which accepts one string parameter.

2. The parameter's value can be any javascript code.

3. `custom.js` contains a function which executes the javascript code passed as parameter value.

4. Any rule invoking `custom.js` from a list outside uBO's official ones or user's own, gets discarded.

That's it. Something like this might achieve what I am proposing? Just my immediate thoughts as a layman who's still trying to figure out scriptlets.

**Comment by uBlock-user** [comment]
Sure it can work out, it's not impossible so to speak, but it's upto @gorhill.

**Comment by jspenguin2017** [comment]
@in4u That wouldn't violate Firefox extension store policy.

**Comment by jspenguin2017** [comment]
Adguard releases a limited version for Firefox, which have all script inject rules hard coded. 

I think the user can still add their own rules, but it is really awkward to use. If you want to inject your own script, try Tampermonkey or Violentmonkey.

**Comment by in4u** [comment]
Interesting. Anyways, one thing I still can't wrap my head around is how what I suggested violates a policy when it's no different from what already exists?

Based on whatever little I understand, in ELI5 terms, scriptlet is like a function with parameters. There are already many such functions which uBO currently uses. I just added a new one. Now either this whole scriptlet thingy should already be in violation of that policy or else adding one more shouldn't be violating any policy. What am I missing?!

**Comment by jspenguin2017** [comment]
A scriptlet is like a function with parameters *that is safe no matter what parameters are passed in*.

**Comment by in4u** [comment]
Thanks for that clarification! Can't this be tackled somehow? Like instead of passing exact javascript code as parameter, encapsulate it in some block so that it just becomes harmless text parameter on its own, e.g. `UBO# [actual javascript code] #UBO` or encode it or something.

**Comment by uBlock-user** [comment]
> By the way, does the Firefox team manually check and approve every single scriptlet?

No, their approval system is now automated like Chrome's Web Store's, Firefox doesn't want any extension which can cause code execution via a file which is part of the extension and gets updated remotely via an external server anytime, this is their AMO policy, this only applies to Firefox stable builds, not dev builds, as dev builds are hosted directly on github, so AMO is bypassed if you're using a dev build on Firefox.

**Comment by in4u** [comment]
Okay, let's leave it for @gorhill to decide the best course of action. At this point, we don't even know if he's at least open to the idea of implementing such a feature. This issue could very well be stamped with a nice big red decline and trashed away as we speak, haha!

**Comment by jspenguin2017** [comment]
> encapsulate it in some block so that it just becomes harmless text parameter on its own

How do you plan to use the "harmless text"? It would just become an expensive no-op machine. 

> Am wondering what the definition of safe as per that policy is. 

Their policy is written in very generic language, honestly no one knows. But executing arbitrary code is probably not safe enough.

**Comment by in4u** [comment]
You're right. Actually, there's no need to encapsulate or sugarcoat the parameter at all. Parameter on its own is harmless anyway. Like, javascript code can be passed as parameter for any of the existing scriptlets too. How the function deals with it is what matters.

Let's say the policy doesn't allow X, Y and Z. The function can then parse the parameter to check for validity and only proceed if X/Y/Z aren't included or applicable, otherwise just drop the rule.

I don't know the technical nitty-gritty. You guys are the expert. Brainstorm together and figure things out. As they say, where there's a will there's a way. Right now things are probably stuck at the will stage, so the question of finding a way doesn't even arise!

**Comment by in4u** [comment]
Um, I meant figuring out a way to handle this feature request and not necessarily beat Google engineers. Adguard has their way of implementing this. You guys can probably think of something else. Worst case, go about it their way. Or, don't. Just create scriptlets which solve relevant issues. As long as the issues can be resolved via official filter lists instead of personal hacks, who even cares about javascript?!

**Comment by jspenguin2017** [comment]
Gorhill doesn't want to bloat the resource database with annoyance filters. 

If you just want to write cookies with filters, which would solve the issue you linked in the opening, [this](https://github.com/NanoAdblocker/NanoDoc/blob/master/en/AdvancedSettings.MD#nanomakeuserfiltersprivileged-switch) and [this](https://github.com/NanoAdblocker/NanoFilters/blob/401f2def1d78475e904c8650372c395bede7c949/NanoFiltersSource/NanoResources.txt#L316) are what you're looking for.

**Comment by gorhill** [comment]
Duplicate of <https://github.com/gorhill/uBlock/issues/3307>.

**Comment by krystian3w** [comment]
#2229
